storage/pruning/pruningStorer.go

```diff
@@ -0,0 +1,444 @@
+	if !exists {
+		return nil, fmt.Errorf("key %s not found in %s",
+			base64.StdEncoding.EncodeToString(key), ps.identifier)
+	}
+
+	if !persisterData.isClosed {
+		return persisterData.persister.Get(key)
+	} else {
+		persister, err := ps.persisterFactory.Create(persisterData.path)
+		if err != nil {
+			log.Debug("open old persister", "error", err.Error())
+			return nil, err
+		}
+
+		err = persister.Init()
+		if err != nil {
+			log.Debug("init old persister", "error", err.Error())
+			return nil, err
+		}
+
+		res, errToRet := persister.Get(key)
+
+		err = persister.Close()
+		if err != nil {
+			log.Debug("close old persister", "error", err.Error())
+			return nil, err
+		}
+
+		if errToRet == nil {
+			return res, nil
+		}
+		log.Warn("get from closed persister",
+			"id", ps.identifier,
+			"epoch", epoch,
+			"key", key,
+			"error", errToRet.Error())
+
+		return nil, fmt.Errorf("key %s not found in %s",
+			base64.StdEncoding.EncodeToString(key), ps.identifier)
+	}
+}
+
+// Has checks if the key is in the Unit.
+// It first checks the cache. If it is not found, it checks the bloom filter
+// and if present it checks the db
+func (ps *PruningStorer) Has(key []byte) error {
+	ps.lock.RLock()
+	defer ps.lock.RUnlock()
+
+	has := ps.cacher.Has(key)
+	if has {
+		return nil
+	}
+
+	if ps.bloomFilter == nil || ps.bloomFilter.MayContain(key) {
+		for _, persister := range ps.activePersisters {
+			if persister.persister.Has(key) != nil {
+				continue
+			}
+
+			return nil
+		}
+	}
+
+	return storage.ErrKeyNotFound
+}
+
+// HasInEpoch checks if the key is in the Unit in a given epoch.
+// It first checks the cache. If it is not found, it checks the bloom filter
+// and if present it checks the db
+func (ps *PruningStorer) HasInEpoch(key []byte, epoch uint32) error {
+	// TODO: this will be used when requesting from resolvers
+	ps.lock.RLock()
+	defer ps.lock.RUnlock()
+
+	has := ps.cacher.Has(key)
+	if has {
+		return nil
+	}
+
+	if ps.bloomFilter == nil || ps.bloomFilter.MayContain(key) {
+		persister, ok := ps.persistersMapByEpoch[epoch]
+		if !ok {
+			return storage.ErrKeyNotFound
+		}
+
+		return persister.persister.Has(key)
+	}
+
+	return storage.ErrKeyNotFound
+}
+
+// Remove removes the data associated to the given key from both cache and persistence medium
+func (ps *PruningStorer) Remove(key []byte) error {
+	ps.lock.Lock()
+	defer ps.lock.Unlock()
+
+	ps.cacher.Remove(key)
+	return ps.activePersisters[0].persister.Remove(key)
+}
+
+// ClearCache cleans up the entire cache
+func (ps *PruningStorer) ClearCache() {
+	ps.cacher.Clear()
+}
+
+// DestroyUnit cleans up the bloom filter, the cache, and the dbs
+func (ps *PruningStorer) DestroyUnit() error {
+	ps.lock.Lock()
+	defer ps.lock.Unlock()
+
+	if ps.bloomFilter != nil {
+		ps.bloomFilter.Clear()
+	}
+
+	ps.cacher.Clear()
+
+	var err error
+	numOfPersistersRemoved := 0
+	totalNumOfPersisters := len(ps.activePersisters)
+	for _, persister := range ps.activePersisters {
```

Comment:
what about the non active persisters?
do you destroy them

Reply:
missed that.. now I iterate over the map of persisters by epoch because it contains all the persisters